### PR TITLE
Make the implementation lazily evaluated.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
As the `ToDo:` recommends.

### CODE REVIEW

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb [hunk 1]
```diff
@@ -14,39 +14,29 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Progression
...
             Dim nodes = New Stack(Of SyntaxNode)()
-
-            Dim result = New List(Of SyntaxNode)
-
             nodes.Push(root)
 
             While nodes.Count > 0
-                cancellationToken.ThrowIfCancellationRequested()
+
+                If cancellationToken.IsCancellationRequested Then Continue While
```
Comment (gafter) [inline src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: Looks like an infinite loop to me.

Comment (AdamSpeight2008) [reply src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: #Done

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb [hunk 2]
```diff
@@ -14,39 +14,31 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Progression
     Partial Friend Class VisualBasicProgressionLanguageService
         Implements IProgressionLanguageService
 
-        Public Function GetTopLevelNodesFromDocument(root As SyntaxNode, cancellationToken As CancellationToken) As IEnumerable(Of SyntaxNode) Implements IProgressionLanguageService.GetTopLevelNodesFromDocument
-            ' TODO: Implement this lazily like in C#?
+        Public Iterator Function GetTopLevelNodesFromDocument(root As SyntaxNode, cancellationToken As CancellationToken) As IEnumerable(Of SyntaxNode) Implements IProgressionLanguageService.GetTopLevelNodesFromDocument
+            If cancellationToken.IsCancellationRequested Then Exit Function
```
Comment (sharwell) [inline src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: ❗️ This is a problem. It requires that the caller always check the cancellation at the end of iteration to ensure it's not cancelled. The return value of a method needs to reveal the validity of the result in cases where cancellation does not produce an exception.

The easiest fix is restoring the original `ThrowIfCancellationRequested` behavior.

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb [hunk 3]
```diff
@@ -14,39 +14,31 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Progression
...
+                    Case Else
+                        For Each child In node.ChildNodes()
+                            nodes.Push(child)
+                        Next
+                End Select
             End While
-
-            Return result
+            nodes.Clear()
+            nodes = Nothing
```
Comment (sharwell) [inline src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: 💭 I don't understand why it clears the stack. Was it meant to be using a pooled object?

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb [hunk 4]
```diff
@@ -14,39 +14,31 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Progression
...
+                If cancellationToken.IsCancellationRequested Then Continue While
+
+                Select Case node.Kind
+                    Case SyntaxKind.ClassBlock, SyntaxKind.DelegateFunctionStatement, SyntaxKind.DelegateSubStatement,
+                         SyntaxKind.EnumBlock, SyntaxKind.ModuleBlock, SyntaxKind.InterfaceBlock,
+                         SyntaxKind.StructureBlock, SyntaxKind.FieldDeclaration, SyntaxKind.SubBlock,
+                         SyntaxKind.FunctionBlock, SyntaxKind.PropertyBlock
+                        Yield node
+                    Case Else
+                        For Each child In node.ChildNodes()
```
Comment (sharwell) [inline src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: 📝 This is exactly the same as the old version of this function, but note forward iteration of the children while using `Stack<T>` as the temporary data storage results in a strange ordering of the result. No need to change it but something to be aware of.

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb [hunk 5]
```diff
@@ -14,39 +14,31 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Progression
...
-                    result.Add(node)
-                Else
-                    For Each child In node.ChildNodes()
-                        nodes.Push(child)
-                    Next
-                End If
+                If cancellationToken.IsCancellationRequested Then Continue While
+
+                Select Case node.Kind
+                    Case SyntaxKind.ClassBlock, SyntaxKind.DelegateFunctionStatement, SyntaxKind.DelegateSubStatement,
```
Comment (sharwell) [inline src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: 💡 One item per line would help readability substantially. In the current form, it takes a long time to answer the question "does this list contain *X*".

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb (hunk 2 above)
Comment (AdamSpeight2008) [reply src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: I'm following the [C# Implementation](https://github.com/dotnet/roslyn/blob/6847f1e5a909395aae9456e8f366cbf4deb86b69/src/VisualStudio/CSharp/Impl/Progression/CSharpProgressionLanguageService.cs#L46), with the check un-negated to enable early exiting.

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb (hunk 3 above)
Comment (AdamSpeight2008) [reply src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: I'll remove this as it is an artefact from now not exit the `while` early.

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb (hunk 5 above)
Comment (AdamSpeight2008) [reply src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: Will do

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: do we use "then exit function" anywhere else in the IDE?  I'd prefer us not use random non-common styles like this.

Comment (AdamSpeight2008) [reply src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: @CyrusNajmabadi [We do](https://github.com/dotnet/roslyn/blob/160c4098fa90cff36fadad616eab2ae967ad2563/src/VisualStudio/VisualBasic/Impl/CodeModel/VisualBasicCodeModelService.vb#L492). There maybe use of the form `... Then Return`, but searching the codebase for that example is tricky. *For me, `Exit Function` makes more sense, as we're not "returning" anything.

File: src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb [hunk 6]
```diff
@@ -14,39 +14,36 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Progression
...
-                    node.Kind = SyntaxKind.SubBlock OrElse
-                    node.Kind = SyntaxKind.FunctionBlock OrElse
-                    node.Kind = SyntaxKind.PropertyBlock Then
-                    result.Add(node)
-                Else
-                    For Each child In node.ChildNodes()
-                        nodes.Push(child)
-                    Next
-                End If
+                If cancellationToken.IsCancellationRequested Then Continue While
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/VisualBasic/Impl/Progression/VisualBasicProgressionLanguageService.vb]: @jasonmalinowski It was me, but I wasn't happy with the impact on clarity and I abandoned the change.

**Comment by AdamSpeight2008** [comment]
@sharwell What would the return value be in valid successful completion and cancellation? 
I don't know personnaly a way to test "inject" a cancellation, so can't add a unit-test to validate the return values. Or if the return value actually has a contextual meaningful information, in either situation.

**Pull request closed without merging**